source/grid/manifold_lib.cc

```diff
@@ -25,6 +25,67 @@
 
 DEAL_II_NAMESPACE_OPEN
 
+
+namespace internal
+{
+  Tensor<1,3> RotateUnitInDirection ( const Tensor<1,3> &u, const Tensor<1,3> &dir)
+  {
+    double theta = dir.norm();
+    if ( theta==0. )
+      {
+        return u;
+      }
+    else
+      {
+        double costheta = cos(theta);
+        double sintheta = sin(theta);
+        Tensor<1,3> dirUnit = dir/theta;
+        return costheta*u + sintheta*dirUnit;
+      }
+  }
+
+
+  // Returns the projection of u onto the plane perpendicular to the unit vector v
+  //    This one is more stable when u and v are nearly equal.
+  Tensor<1,3> ProjectPerpUnitDiff ( const Tensor<1,3> &u, const Tensor<1,3> &v)
+  {
+    Tensor<1,3> ans = u-v;
+    ans -= (ans*v)*v;
+    return ans;             // ans = (u-v) - ((u-v)*v)*v
+  }
+
+  // helper function to compute a vector orthogonal to a given one.
+  Point<3>
+  compute_normal(const Tensor<1,3> &vector)
+  {
+    AssertThrow(vector.norm() != 0.,
```

Comment:
Since you only use this for testing for a non-zero vector, I suggest to use `vector.norm_square() != 0`.

Reply:
should not that be `Assert`?